pkg/api/api.go

```diff
@@ -0,0 +1,229 @@
+	return code == statusAPIError || code == http.StatusBadRequest
+}
+
+// queryResult contains result data for a query.
+type queryResult struct {
+	Type   model.ValueType `json:"resultType"`
+	Result interface{}     `json:"result"`
+
+	// The decoded value.
+	v model.Value
+}
+
+func (qr *queryResult) UnmarshalJSON(b []byte) error {
+	v := struct {
+		Type   model.ValueType `json:"resultType"`
+		Result json.RawMessage `json:"result"`
+	}{}
+
+	err := json.Unmarshal(b, &v)
+	if err != nil {
+		return err
+	}
+
+	switch v.Type {
+	case model.ValScalar:
+		var sv model.Scalar
+		err = json.Unmarshal(v.Result, &sv)
+		qr.v = &sv
+
+	case model.ValVector:
+		var vv model.Vector
+		err = json.Unmarshal(v.Result, &vv)
+		qr.v = vv
+
+	case model.ValMatrix:
+		var mv model.Matrix
+		err = json.Unmarshal(v.Result, &mv)
+		qr.v = mv
+
+	default:
+		err = fmt.Errorf("unexpected value type %q", v.Type)
+	}
+	return err
+}
+
+type apiResponse struct {
+	Status    string              `json:"status"`
+	Data      json.RawMessage     `json:"data"`
+	ErrorType promapiv1.ErrorType `json:"errorType"`
+	Error     string              `json:"error"`
+	Warnings  []string            `json:"warnings,omitempty"`
+}
+
+func do(ctx context.Context, client promapi.Client, req *http.Request) (*http.Response, []byte, promapiv1.Warnings, error) {
+	resp, body, err := client.Do(ctx, req)
+	if err != nil {
+		return resp, body, nil, err
+	}
+
+	code := resp.StatusCode
+
+	if code/100 != 2 && !apiError(code) {
+		errorType, errorMsg := errorTypeAndMsgFor(resp)
+		return resp, body, nil, &promapiv1.Error{
+			Type:   errorType,
+			Msg:    errorMsg,
+			Detail: string(body),
+		}
+	}
+
+	var result apiResponse
+
+	if http.StatusNoContent != code {
+		if jsonErr := json.Unmarshal(body, &result); jsonErr != nil {
+			return resp, body, nil, &promapiv1.Error{
+				Type: ErrBadResponse,
+				Msg:  jsonErr.Error(),
+			}
+		}
+	}
+
+	if apiError(code) && result.Status == "success" {
+		err = &promapiv1.Error{
+			Type: ErrBadResponse,
+			Msg:  "inconsistent body for response code",
+		}
+	}
+
+	if result.Status == "error" {
+		err = &promapiv1.Error{
+			Type: result.ErrorType,
+			Msg:  result.Error,
+		}
+	}
+
+	return resp, result.Data, result.Warnings, err
+}
+
+// doGetFallback will attempt to do the request as-is, and on a 405 it will fallback to a GET request.
+func doGetFallback(
+	ctx context.Context,
+	client promapi.Client,
+	u *url.URL,
+	args url.Values,
+	cache bool,
+) (*http.Response, []byte, promapiv1.Warnings, error) { //nolint:unparam
+	req, err := http.NewRequest(http.MethodPost, u.String(), strings.NewReader(args.Encode()))
+	if err != nil {
+		return nil, nil, nil, err
+	}
+	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
+	if !cache {
+		req.Header.Set("Cache-Control", "no-store")
+	}
+
+	resp, body, warnings, err := do(ctx, client, req)
+	if resp != nil && resp.StatusCode == http.StatusMethodNotAllowed {
```

Comment:
I think we can simplify this code much more. Let's remove else and use `do` explicitly if method not allowed (: 

Also we have to close and exhaust body in both cases (:

Reply:
Updated.